### PR TITLE
feat: improve Parity class

### PR DESCRIPTION
**Pull request by redeboer**
I had some problems when using the `ParticleCollection` in some external notebook when doing some list comprehensions involving parity. This PR makes the `Parity` class behave more as expected.

### CODE REVIEW

File: expertsystem/data.py [hunk 1]
```diff
@@ -32,12 +33,29 @@ def __eq__(self, other: object) -> bool:
...
+
     def __int__(self) -> int:
         return self.value
 
     def __neg__(self) -> "Parity":
         return Parity(-self.value)
 
+    def __hash__(self) -> int:
+        return self.__value
+
```
Comment (redeboer) [inline expertsystem/data.py:58]: Side note: I had problems with the (previously) missing `__hash__` method when creating a `set` of `tuples` with particle info (in this case a set of tuples of parities with spins).

Comment (spflueger) [reply expertsystem/data.py:58]: I can't really follow the need for this functionality?

File: expertsystem/data.py [hunk 2]
```diff
@@ -32,12 +33,29 @@ def __eq__(self, other: object) -> bool:
...
+    def __ge__(self, other: Any) -> bool:
+        return self > other or self == other
+
+    def __lt__(self, other: Any) -> bool:
+        return not self >= other
+
+    def __le__(self, other: Any) -> bool:
+        return not self > other
+
```
Comment (spflueger) [inline expertsystem/data.py:49]: why is this needed?

File: expertsystem/data.py (hunk 1 above)
Comment (redeboer) [reply expertsystem/data.py:58]: I tried to do something like this:
```python
>>> from expertsystem import io
p.parity, p.spin) for p in pdg.values()}
>>> pdg = io.load_pdg()
>>> {(p.parity, p.spin) for p in pdg.values()}
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 1, in <setcomp>
TypeError: unhashable type: 'Parity'
```

File: expertsystem/data.py (hunk 2 above)
Comment (redeboer) [reply expertsystem/data.py:49]: Just to have expected behaviour like:
```python
from expertsystem import io
pdg = io.load_pdg()
{p for p in pdg if p.parity < 0}
```
One could wonder about the `<=` etc but I would have that as well if you introduce this '`int`-like behaviour.

File: expertsystem/data.py (hunk 1 above)
Comment (spflueger) [reply expertsystem/data.py:58]: Ok, I'm not sure what the use case for doing something like this would be. I guess a user wants to reduce the particle to some important qns. Btw, you could still do `(p.parity.value,...`

File: expertsystem/data.py (hunk 2 above)
Comment (spflueger) [reply expertsystem/data.py:49]: I just don't see a use case when someone wants to do such an operation.

File: expertsystem/data.py (hunk 1 above)
Comment (redeboer) [reply expertsystem/data.py:58]: That is true. Or, alternatively, `int(p.parity)`.

File: expertsystem/data.py (hunk 2 above)
Comment (redeboer) [reply expertsystem/data.py:49]: The `ParticleCollection.filter` function has become a bit of a quantum number search engine (see [interactive exampels](https://pwa.readthedocs.io/projects/expertsystem/en/latest/interactive.html#particle-database)). So I want that function to work as conveniently as possible.
```python
pdg.filter(lambda p: p.parity < 0 and p.spin == 1)
```
But I'm aware that there is a risk: when does convenience become unexpected behaviour?

File: expertsystem/data.py (hunk 1 above)
Comment (redeboer) [reply expertsystem/data.py:58]: See also https://github.com/ComPWA/expertsystem/pull/271#discussion_r494829408

File: expertsystem/data.py (hunk 2 above)
Comment (spflueger) [reply expertsystem/data.py:49]: Ok thats maybe a bad example as this would be the same as `p.parity == -1`, but I see your point now.
Like if you want unknown and -1 parity you could do <= 0 or < 1 instead of `parity == -1 and parity == 0` (or None). And its a bit more convenient than having to write `parity.value`

Comment (redeboer) [reply expertsystem/data.py:49]: This point is however related to #259, which would make `Parity` a more complicated object. Right now, it's just a 'read-only' `int` that can only be ±1.

File: expertsystem/data.py [hunk 3]
```diff
@@ -32,12 +33,29 @@ def __eq__(self, other: object) -> bool:
...
+    def __gt__(self, other: Any) -> bool:
+        if isinstance(other, Parity):
+            return self.value > other.value
+        return self.value > int(other)
+
+    def __ge__(self, other: Any) -> bool:
+        return self > other or self == other
+
+    def __lt__(self, other: Any) -> bool:
+        return not self >= other
```
Comment (spflueger) [inline expertsystem/data.py:45]: im not sure about standard python procedure here, but I'm quite used to just swapping the operands, so here
`self < other` -> `other > self`

Comment (redeboer) [reply expertsystem/data.py:45]: You mean `self < other` instead of `not self >= other`? But that would cause an infinite recursion

Comment (spflueger) [reply expertsystem/data.py:45]: No so the general idea is you define either `<` or `>`, and then you just switch the arguments if you want the opposite.
So if you defined greater than (>) then you would define less than (A < B) as `return B > A`.

Comment (redeboer) [reply expertsystem/data.py:45]: Not sure what you mean. The methods implemented here were required to get the following tests working:
https://github.com/ComPWA/expertsystem/blob/148b9419d5001f5f18447ea2346699052c2d33e7/tests/unit/test_data.py#L81-L89

Comment (spflueger) [reply expertsystem/data.py:45]: I'm saying
```python3
def __lt__(self, other: Any) -> bool:
        return other > self
```
looks better than
```python3
def __lt__(self, other: Any) -> bool:
        return not self >= other
```

Comment (redeboer) [reply expertsystem/data.py:45]: Ah ok. But still, it causes a max recursion:
```
expertsystem/data.py:45: in __lt__
    return other > self
expertsystem/data.py:45: in __lt__
    return other > self
E   RecursionError: maximum recursion depth exceeded in comparison
!!! Recursion detected (same locals & position)
```
(this is done on 148b9419 with the exact replacement as you describe)

Comment (spflueger) [reply expertsystem/data.py:45]: Does not make sense to me. I would say you made a mistake in your implementation then. So you should define 
```python3
def __gt__(self, other: Any) -> bool:
        return self.value > other.value 
```
(Note i simplified the implementation)
So not keep pointing one function to another. So you define one function and make all the other ones point to this one

Comment (redeboer) [reply expertsystem/data.py]: ca40fb25

Comment (spflueger) [reply expertsystem/data.py]: Ok just for clarification: the reason for the endless loop was the comparison for an `int` with a `Parity`

**Review by spflueger — APPROVED** [review]
